### PR TITLE
Add distress signals

### PR DESCRIPTION
**Pull request by Samsterious**
## About the PR
This PR replaces the pre-existing station events with distress signals, a type of station event that provides mappers a tool with which to inject non-invasive gameplay opportunities into the round with minimal effort and a great deal of flexibility.

The essential gameplay loop of a distress signal is the arrival of a shuttle in distress for one reason or another. Perhaps there are injured crew in need of rescue, hostiles onboard that need to be detained, a plasma fire trapping some of the occupants on the bridge, or the engine got blasted apart by an asteroid. If players react quickly enough to identify the problem, address the hazards, and reverse the damage, the distress signal will be cleared and the grid will leave the map.

Mappers may make use of objective spawners, which create relevant objects pre-configured to register themselves with the distress signal when the grid is loaded. For instance, an Injured spawner will create a randomly-generated NPC configured as a patient that players must rescue. A BadAir spawner will create an air sensor that must read a safe atmosphere. A NoPower spawner will create an APC that must be recharged. A Hostile spawner will create a gun-toting baddie that needs to be detained.

The best distress signal grids will contain few objectives but numerous problems in the way to completing those objectives, giving players the opportunity to address the situation in their own way. If the shuttle was spaced, but all its critical components are intact, and the only objective is to rescue the injured crew member, they may return the crew member to their shuttle in an EVA suit and call it a day, or go to the trouble of fixing up their shuttle for them and repressurizing the cabin. If there is a plasma fire aboard, and it has taken out some of the atmospherics hardware, then the players may choose to either drag an air canister in and open the valve, or actually repair the shuttle's atmos setup.

Under the hood, there are five moving parts involved:
- The `DistressSignalRule` is a system which manages all active events of this type at once.
- `DistressSignalRuleComponent`s store data specific to each individual instance of the event.
- `DistressSignalObjectiveComponent`s and their extending classes (such as `DistressSignalInjuredComponent`) track data relating to specific objectives.
- The `DistressSignalObjectiveSystem` and its extending classes manage all of the corresponding components, providing the logic that determines whether a specific objective has been completed.
- Spawner prototypes, such as `RandomHumanoidSpawnerDistressInjured`, provide mappers a quick tool with which to stamp out objectives onto a distress signal grid.

**Changelog**
:cl: Fireheart&
- add: Distress signal events will now periodically appear on the map.

### CODE REVIEW

File: Content.Server/_Lagrange/StationEvents/Events/DistressSignalRule.cs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+namespace Content.Server.StationEvents.Events;
+public sealed class DistressSignalRule : StationEventSystem<DistressSignalRuleComponent>
+{
+    [Dependency] private readonly IMapManager _mapManager = default!;
+    [Dependency] private readonly MapLoaderSystem _map = default!;
+    [Dependency] private readonly ShuttleSystem _shuttle = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+
+    private List<(Entity<TransformComponent> Entity, EntityUid MapUid, Vector2 LocalPosition)> _playerMobs = new();
```
Comment (makyo) [inline Content.Server/_Lagrange/StationEvents/Events/DistressSignalRule.cs:26]: For readability's sake, can we get a newline here?

File: Content.Server/_Lagrange/StationEvents/Events/DistressSignalRule.cs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            return;
+        }
+
+        var shuttleMap = _mapManager.CreateMap();
+        var options = new MapLoadOptions
+        {
+            LoadMap = true,
+        };
+
+        if (!_map.TryLoad(shuttleMap, shuttleProto.MapPath.ToString(), out var gridUids, options))
```
Comment (makyo) [inline Content.Server/_Lagrange/StationEvents/Events/DistressSignalRule.cs:48]: If we are returning here, do we need to `GameTicker.EndGameRule(uid);`?

File: Content.Server/_Lagrange/StationEvents/Events/DistressSignalRule.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            return;
+        component.GridUid = gridUids[0];
+        if (component.GridUid is not EntityUid gridUid)
```
Comment (makyo) [inline Content.Server/_Lagrange/StationEvents/Events/DistressSignalRule.cs:55]: Ditto

File: Content.Server/_Lagrange/StationEvents/Events/DistressSignalRule.cs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            }
+        }
+
+        if (successCount + failureCount >= component.Objectives.Count)
+        {
+            component.ObjectivesCompleted = successCount > 0;
+
+            // Ensure the objectives *stay* completed.
+            component.TimerRunning = true;
+            Timer.Spawn(TimeSpan.FromSeconds(15), () =>
```
Comment (makyo) [inline Content.Server/_Lagrange/StationEvents/Events/DistressSignalRule.cs]: You would know better than I, but is this worth keeping in config? Like, if the spawning process takes shorter or longer than 15s, can we adjust that per deploy? If it's constant, cool.

File: Resources/Prototypes/GameRules/events.yml [hunk 5]
```diff
@@ -428,17 +428,17 @@
...
+#   components:
+#   - type: StationEvent
+#     weight: 5
+#     earliestStart: 20
+#     reoccurrenceDelay: 60
+#     startAnnouncement: station-event-ion-storm-start-announcement
+#     startAudio:
+#       path: /Audio/Announcements/ion_storm.ogg
+#     duration: 1
+#   - type: IonStormRule
```
Comment (makyo) [inline Resources/Prototypes/GameRules/events.yml:444]: While I think I might have liked this in a separate branch to reference later, it's also so easy to play with, it's not worth it. :+1:

**Review by makyo — COMMENTED** [review]
LGTM with trivials, I think.

File: Resources/Prototypes/GameRules/events.yml (hunk 5 above)
Comment (Samsterious) [reply Resources/Prototypes/GameRules/events.yml:444]: My thinking was that this is in-scope for this PR on account of the fact that distress signals essentially replace the vanilla events system entirely (most of them are just annoying and do not actually offer any interesting gameplay).

Comment (Samsterious) [reply Resources/Prototypes/GameRules/events.yml:444]: That said, ion storm good.

**Pull request merged**